Custom catalogs from `IcebergSource`

rymurr opened the pull request:
I wanted to start a conversation about how to make `IcebergSource` compatible with custom Iceberg Catalogs in both Spark 2&3.

This first attempt is meant to give a straw man for how this could be done. It uses code from `LookupCatalog` to fetch a Catalog and Identifier from a path. We then try and extract an Iceberg table from it.

This works by specifying the catalog(`iceberg_catalog`) below in the path or by setting a default catalog in spark config.
`df.write.format("iceberg").mode("append").save("iceberg_catalog.testing.foo")`

Several problems:
1) still have the fragile path check to delegate to the Hadoop catalog
2) not strictly backwards compatible one has to specify a catalog as part of the path or a default catalog
3) not 100% sure this will work for Spark2
4) Copies code from `LookupCatalog` it would be nice to use that (or similar) directly but that would involve putting scala code in the code path.
5) Hard to pass parameters to the catalog as we do now

I am soliciting opinions on what might be a better way of doing custom catalogs, has anyone else been thinking about this?

cc @rdblue @jacques-n @laurentgo

jackye1995 commented:
Why not do something like the following, which can make the whole thing more backwards compatible:

```java
df.write.format("iceberg")
  .mode("append")
  .option("catalog-impl", "com.my.own.CatalogImpl")
  .save("testing.foo")
```

rymurr commented:
> Why not do something like the following, which can make the whole thing more backwards compatible:

I was thinking similar but as per #1587 we don't want to hold catalog configs in the hadoop `Configuration`. Logically we would then put the catalog config into `option`s. To me this leads to verbose/repetitive `read`/`write` statements (see below). I am looking for a solution that still enables session wide catalog configs to be set and this [comment](https://github.com/apache/iceberg/pull/1587#discussion_r521614936) implied that it is technically possible.

```
df.write.format("iceberg")
  .mode("append")
  .option("catalog-impl", "com.my.own.CatalogImpl")
  .option("catalog-option1", "value1")
  .option("catalog-option2", "value2")
  ...
  .save("testing.foo")
```

rymurr commented:
@jackye1995 I think I misunderstood your original comment. When you said:

```
df.write.format("iceberg")
  .mode("append")
  .option("catalog-impl", "com.my.own.CatalogImpl")
  .save("testing.foo")
```

Did you mean then use `catalog-impl` parameter to look up an existing iceberg catalog by type? Which (hopefully) would have been constructed previously, so we wouldn't have to set extra options like I mentioned above. Does that sound right? While much simpler I think its important that the user doesn't have to set options in the `read`/`write` spark commands, rather set them once in `SparkConf`.

I wonder if we could set a default catalog in `SparkConf`` (eg `spark.iceberg.default_catalog=com.my.own.CatalogImpl`)? If set the readers/writers will use that unless an `option` is passed. If neither is passed it would default to the current Hive/HDFS impl.

I believe that this still causes an issue in Spark2 as it doesn't have the same (spark) catalog support. 

Thoughts?

rdblue commented:
I agree that we will need to mimic the behavior of `LookupCatalog` and use the catalog manager. I think we can fix some of the problems with this by using the current catalog from the catalog manager if there is no catalog identifier in the table name.

Here's what I think is the _correct_ thing to do for table identifiers in Spark 3:
1. Parse the identifier into parts
2. Follow the logic in `LookupCatalog`: if there is only one part, use the current catalog and current namespace. If there are multiple parts, check whether the first one is a catalog and use it if it is. After this, we have a catalog and a table identifier.
3. Check that the catalog is an Iceberg catalog. If not, throw an exception. _Not entirely sure about this one._
4. Return the catalog and identifier through [`SupportsCatalogOptions`](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsCatalogOptions.java).

That way, we're just using the source to translate back to catalogs.

The problem with this is the behavior when there is no catalog specified in the identifier. The current behavior is to use a `HiveCatalog` that connects to the URI from `hive-site.xml` like the built-in Spark session catalog. That conflicts with the SQL behavior of using the current catalog, but may be reasonable to keep compatibility. But then the problem is that we may not have a registered Iceberg catalog that uses that URI. If not, then there is no catalog for the source to delegate to and we need to create one because a catalog is required if the source implements `SupportsCatalogOptions`.

Another option is to change the current behavior slightly and go with the "correct" logic for Spark 3 and delegate to the current catalog. That would mean we always have a catalog to delegate to without creating one. But the trade-off is that when the current catalog has changed, the table loaded by `IcebergSource` would change, too. I'd be open to this option.

Last, how to handle path URIs: I talked with Anton and Russell about this yesterday and we think that we should make it so that every Spark catalog can load special path identifiers, just like we do today in `IcebergSource`. To do this, we would need to have a way to pass an identifier that signals this behavior back from `SupportsCatalogOptions`, like `iceberg.hdfs://nn:8020/path/to/table`. Then detect those and return Hadoop tables for them.

What do you think?

aokolnychyi commented:
A couple of questions before we try to come up with a solution.

1. How important or even valid is it to load non-path based tables through our source in Spark 3? We decided to interpret `path` option as a table identifier in 2.4 to avoid limitations of the DS V2 API at that point. Since we have proper catalog API in Spark 3, shall we assume all catalog-based tables should be loaded through it and not through source? We've replicated the `LookupCatalog` logic in Spark3Util but I am not sure it is a good idea to use that in the source.

2. It is a bit unclear to me how we are going to produce the catalog name from `SupportsCatalogOptions`. The catalog manager does not provide a list of catalogs that were registered. Will we require a catalog name in options?

rdblue commented:
> How important or even valid is it to load non-path based tables through our source in Spark 3?

The purpose is to maintain compatibility with 2.4. If you have a working job that used IcebergSource to load a table from the Hive catalog, then that should continue to work.

> It is a bit unclear to me how we are going to produce the catalog name from SupportsCatalogOptions.

The same way that we lookup catalogs in Spark: If the first part is a catalog, return it. Otherwise, use a default.

The problem with this is how we want to maintain compatibility with Spark 2.4. In 2.4, the behavior is that the default catalog is the Hive session catalog and that can't be changed. In 3, I think there is a good argument that since the session catalog is the default unless you configure a different one (spark.sql.defaultCatalog) then we can respect the default catalog.

aokolnychyi commented:
> Last, how to handle path URIs: I talked with Anton and Russell about this yesterday and we think that we should make it so that every Spark catalog can load special path identifiers, just like we do today in IcebergSource. To do this, we would need to have a way to pass an identifier that signals this behavior back from SupportsCatalogOptions, like iceberg.hdfs://nn:8020/path/to/table. Then detect those and return Hadoop tables for them.

+1 to that. I think we have enough info to get #1306 done. While it definitely makes sense to accept `iceberg.path` in the session catalog, shall we also allow just `path` alone in non-session catalogs? Otherwise, it is going to be weird if the catalog name is iceberg too. The identifier will be `iceberg.iceberg.path` then.

aokolnychyi commented:
> The same way that we lookup catalogs in Spark: If the first part is a catalog, return it. Otherwise, use a default.

I like this approach but will it be backward compatible if the user does not set the session catalog as Iceberg session catalog?

rdblue commented:
> I like this approach but will it be backward compatible if the user does not set the session catalog as Iceberg session catalog?

It's hard to define because we would be respecting a setting that wasn't present in 2.4. I think it's a reasonable thing to do as long as no catalog setting changes maintain compatibility with 2.4. Otherwise, the user is opting for different behavior by changing the default catalog.

rymurr commented:
Thanks all for the input. 

So I think there is enough info that I can clean up this PR to use `SupportsCatalogOptions` and the process described above to extract a table/catalog from the source.

* If there is a catalog in the identifier -> use that catalog
* if no catalog in the identifier -> use session/default catalog (favouring Spark3 convention over IcebergSource in 2.4 convention)
* if catalog in the identifier isn't an iceberg catalog -> throw
* if path then handle with `iceberg.path` option as per above. Seems like should be done as a separate patch?

Does that make sense? Am I missing anything?

PS Any thoughts on custom catalogs in Spark2.4? I don't know what the migration plan looks like but I feel like Spark2.4 will be a thing for some time.

rdblue commented:
> if path then handle with iceberg.path option as per above. Seems like should be done as a separate patch?

I think that the problem earlier was that once you implement `SupportsCatalogOptions`, you can't also return a table. You have to return a catalog and identifier for everything.

I think that will require that all of our catalogs support an identifier that contains a path and use that path to load the table.

> Any thoughts on custom catalogs in Spark2.4?

We could do this in the `IcebergLoader` for 2.4. I think that could be a nice feature.

rymurr commented:
> I think that will require that all of our catalogs support an identifier that contains a path and use that path to load the table.

Agreed. I was going to open a PR for that to merge before this one. Unless you want it all in 1 PR?

rdblue commented:
Separate PRs works for me. I think you'll need to add that first, or else you won't be able to get tests passing.

rymurr commented:
FYI this has been updated based on our discussions last week. Build is still failing but once #1843 is closed it should be good to go.

rdblue commented on spark3/src/test/java/org/apache/iceberg/spark/source/SetupSourceCatalog.java:
```diff
@@ -0,0 +1,41 @@
+    setupSparkCatalog(spark, SparkSessionCatalog.class.getName());
+  }
+
+  public static void setupSparkCatalog(SparkSession spark, String catalogName) {
```
I don't think `catalogName` is accurate because catalogs have names (this one is `spark_catalog`). It should be `catalogClass` to be more clear.

rymurr replied:
:+1:

rdblue commented on spark3/src/test/java/org/apache/iceberg/spark/source/SetupSourceCatalog.java:
```diff
@@ -0,0 +1,41 @@
+import org.apache.iceberg.spark.SparkSessionCatalog;
+import org.apache.spark.sql.SparkSession;
+
+public final class SetupSourceCatalog {
```
Why not add this to the `iceberg-spark` module? The only thing that isn't compatible is `SparkSessionCatalog.class.getName()`. The 2.4 code should ignore these settings, and this change would be a lot smaller.

rymurr replied:
cool, I was thinking similar but wanted to get your opinion first

rdblue commented on spark3/src/main/java/org/apache/iceberg/spark/source/IcebergSource.java:
```diff
@@ -56,48 +61,62 @@ public boolean supportsExternalMetadata() {
-  protected Table findTable(Map<String, String> options, Configuration conf) {
+  private Pair<String, TableIdentifier> tableIdentifier(CaseInsensitiveStringMap options) {
+    CatalogManager catalogManager = SparkSession.active().sessionState().catalogManager();
+    Namespace defaultNamespace = Namespace.of(catalogManager.currentNamespace());
```
@RussellSpitzer, I seem to remember looking at very similar logic recently. Did we commit that anywhere that we can reuse?

rdblue replied:
The logic was added to `Spark3Util`: https://github.com/apache/iceberg/blob/master/spark3/src/main/java/org/apache/iceberg/spark/Spark3Util.java#L607

Can you update this to use those utils?

rymurr replied:
lovely, thats fixed

rymurr commented on spark3/src/test/java/org/apache/iceberg/spark/source/TestSparkSchema3.java:
```diff
@@ -1,23 +0,0 @@
-/*
```
A potential complication: `SupportsCatalogOptions` doesn't allow for specifying the schema. I don't know how much people rely on this feature but it is a breaking change for the IcebergSource

rdblue commented on spark/src/test/java/org/apache/iceberg/TestScanTaskSerialization.java:
```diff
@@ -60,11 +62,17 @@
       optional(3, "c3", Types.StringType.get())
   );
 
+
```
Nit: unnecessary whitespace change.

rymurr replied:
fixed

rdblue commented on spark/src/test/java/org/apache/iceberg/spark/source/SetupSourceCatalog.java:
```diff
@@ -0,0 +1,54 @@
+public final class SetupSourceCatalog {
+
+  private SetupSourceCatalog() {
+
```
Nit: can you remove this newline?

rymurr replied:
fixed

rdblue commented on spark3/src/main/java/org/apache/iceberg/spark/source/IcebergSource.java:
```diff
@@ -56,48 +61,62 @@ public boolean supportsExternalMetadata() {
+    } catch (NoSuchTableException e) {
+      throw new org.apache.iceberg.exceptions.NoSuchTableException(e, "Cannot find table for %s.", ident);
+    }
+    throw new org.apache.iceberg.exceptions.NoSuchTableException("Cannot find table for %s.", ident);
```
I think the correct exception is the Spark exception since this is going to be called from Spark code.

rymurr replied:
The Spark `NoSuchTableException` is typed which would change the interface. It appears there is no well defined way to return from this function w/o a table being found. It NPEs if you return null so an untyped exception seems to be the expected way to denote no table found. I thought the Iceberg `NoSuchTableException` was the best compromise here.

rdblue replied:
Got it, I agree in that case. Thanks for explaining! You may want to add a comment here to explain in the code as well.

rdblue commented on spark3/src/main/java/org/apache/iceberg/spark/source/IcebergSource.java:
```diff
@@ -56,48 +61,62 @@ public boolean supportsExternalMetadata() {
+      ident = scala.collection.JavaConverters.seqAsJavaList(SparkSession.active().sessionState().sqlParser().parseMultipartIdentifier(path));
+    } catch (ParseException e) {
+      try {
+        ident = scala.collection.JavaConverters.seqAsJavaList(SparkSession.active().sessionState().sqlParser().parseMultipartIdentifier(String.format("`%s`", path)));
```
Why is this second attempt done?

Wrapping the identifier in ` should escape the entire string as a single identifier, so the parser would return the original path as one component.

I think this would be equivalent to `ident = path`.

rymurr replied:
you are correct. Fixed

rdblue commented on spark3/src/main/java/org/apache/iceberg/spark/source/IcebergSource.java:
```diff
@@ -56,48 +61,62 @@ public boolean supportsExternalMetadata() {
+  @Override
+  public String extractCatalog(CaseInsensitiveStringMap options) {
+    String catalogName = tableIdentifier(options).first();
+    return (catalogName == null) ? SupportsCatalogOptions.super.extractCatalog(options) : catalogName;
```
The default implementation always returns `spark_catalog`, not the current catalog. Since we want to use the current catalog when it isn't defined in the identifier, `tableIdentifier(options)` should fill it in. That would simplify this logic because it should always be non-null.

rymurr replied:
Ahh, yes. I just re-read the comments. Make sense, fixed.

rdblue commented on spark3/src/main/java/org/apache/iceberg/spark/source/IcebergSource.java:
```diff
@@ -56,48 +61,62 @@ public boolean supportsExternalMetadata() {
+      return Pair.of(null, TableIdentifier.of(defaultNamespace, ident.get(0)));
+    } else if (ident.size() == 2) {
+      if (catalogManager.isCatalogRegistered(ident.get(0))) {
+        return Pair.of(ident.get(0), TableIdentifier.of(defaultNamespace, ident.get(1))); //todo what if path?
```
This shouldn't fill in the default namespace. If the identifier was two parts, like `prod.items`, then it is not correct to modify that to be `prod.default.items`.

I think that this should use the same logic as the `else`.

rymurr replied:
ok, makes sense. That also clarifies the todo i left there. Fixed to not add a default namespace. Does the same logic apply for the case above. If I pass `items` it will get re-written to `default_catalog.default.items`, should it only be `default_catalog.items`?

rdblue commented on spark3/src/main/java/org/apache/iceberg/spark/source/IcebergSource.java:
```diff
@@ -56,48 +61,62 @@ public boolean supportsExternalMetadata() {
+        throw new RuntimeException(e);
+      }
+    }
+    if (ident.size() == 1) {
```
Nit: could you separate the control flow statements with a newline?

rymurr replied:
:+1:

rymurr commented:
Thanks for the comments @rdblue . Think this is close once we get #1804 merged and this build passing :-)

jackye1995 commented on spark3/src/main/java/org/apache/iceberg/spark/source/IcebergSource.java:
```diff
@@ -56,48 +62,62 @@ public boolean supportsExternalMetadata() {
-      return tables.load(path);
+    if (ident.size() == 1) {
+      return Pair.of(currentCatalogName, TableIdentifier.of(defaultNamespace, ident.get(0)));
+    } else if (ident.size() == 2) {
```
Because we are now assigning an empty namespace for `size==2` case,  think it can be merged together with the `else` case. We only get `ident.subList(1, ident.size()).toArray(new String[0])` simplified to `ident.get(1)`, but the rest are all duplicates.

rymurr replied:
done

jackye1995 reviewed (commented):
LGTM, but I think we should have detailed documentation for these evaluation rules, with examples, both in the code and in the website doc for using `IcebergSource`.

rymurr commented on spark3/src/test/java/org/apache/iceberg/spark/source/TestIcebergSource.java:
```diff
@@ -1,27 +1,23 @@
 /*
- * Licensed to the Apache Software Foundation (ASF) under one
```
fixed, dunno how that happened...

rymurr commented:
Thanks for the review @jackye1995, I have added some comments/docs and addressed your comments.

rdblue commented:
> I like this approach but will it be backward compatible if the user does not set the session catalog as Iceberg session catalog?

I think I misinterpreted this comment from @aokolnychyi before. I thought he was talking about default catalogs, but now that I've thought about this a bit more there's still a problem if there is no Iceberg catalog configured in the environment. In 2.4, there is no need to configure any Iceberg catalog and we should support the same case, where a user wants to load a table and hasn't configured any catalogs in 3.0.

I'm not sure how we want to do that exactly. One idea is to ensure that at least one catalog is registered by creating the properties for it. That's a bit ugly, but I can't think of other solutions.

rymurr commented:
> I'm not sure how we want to do that exactly. One idea is to ensure that at least one catalog is registered by creating the properties for it. That's a bit ugly, but I can't think of other solutions.

One option is to default to the Hive Catalog if no other properties are present. The approach I took in #1875 is to try the Hive catalog if all else fails

rdblue commented:
> One option is to default to the Hive Catalog if no other properties are present.

I don't think that would work because some catalog name must be returned. We could only use a hive catalog if we've create one.

rymurr commented:
> I don't think that would work because some catalog name must be returned. We could only use a hive catalog if we've create one.

Ah, yes. I was thinking of it from the `getTable` perspective. I see two options:

1. as you suggested we register a catalog ourselves if none are registered
2. we fail

Option 2 is simple but obviously breaks backward compatibility. I don't know how people are using Spark3 and if catalogs are widely adopted so I don't know how bad of a break it is. Option 1 is as you say rather ugly but is backwards compatible.

My understanding for Option 1 is the following:
* check if the requested catalog (or current catalog) is able to handle Iceberg
* if yes great, else create and register a `HiveCatalog` and replace the requested catalog w/ the created one

This is rather woolly as I don't think we should change the catalog and I don't know how to reliably determine if a catalog supports Iceberg. A `CatalogPlugin` may be a delegating catalog or something that handles Iceberg tables but doesn't extend our `BaseCatalog`.

I suppose option 3 is to proceed but warn that we can't guarantee the user is actually writing an Iceberg table. But who reads log messages ;-)

I will post a straw man for option 1 to at least get something to discuss.

rymurr commented:
The most recent commit will try and use a `HiveCatalog` if the default catalog isn't an iceberg catalog. On one hand it doesn't require changing tests (woohoo!) on the other its not pretty, and not overly reliable. The forced catalog settings aren't quite right either: they turn off caching for example, though perhaps we want that behaviour for the `IcebergSource` as it didn't cache originally.

rdblue commented on spark3/src/main/java/org/apache/iceberg/spark/source/IcebergSource.java:
```diff
@@ -19,22 +19,49 @@
+ * How paths/tables are loaded when using spark.read().format("iceberg").path(table)
+ *
+ *  table = "file:/path/to/table" -> loads a HadoopTable at given path
+ *  table = "catalog.`file:/path/to/table`" -> loads a HadoopTable at given path using settings from 'catalog'
```
What does "using settings from 'catalog'" mean in practice? Hadoop tables are loaded using a generic `HadoopTables`, so the catalog doesn't really affect it at all, I think. In that case, this is promising something that it doesn't need to and I'd prefer to avoid making claims about behavior like that.

rymurr replied:
Yes you are correct. I was thinking of a catalog having different hadoop `Configuration` but I don't think that makes any sense. Fixed now.

rdblue commented on spark3/src/main/java/org/apache/iceberg/spark/source/IcebergSource.java:
```diff
@@ -19,22 +19,49 @@
+ *  table = "xxx.tablename" -> if xxx is a catalog load "tablename" from the specified catalog. Otherwise
+ *          load "xxx.tablename" from current catalog
+ *  table = "xxx.yyy.tablename" -> if xxx is a catalog load "yyy.tablename" from the specified catalog. Otherwise
+ *          load "xxx.yyy.tablename" from current catalog
```
In these examples, I think it is easier to understand if you separate the cases and use `catalog` and `namespace`, rather than "if xxx is a catalog . . . otherwise . . .".

rymurr replied:
agreed, fixed.

rdblue commented on spark3/src/main/java/org/apache/iceberg/spark/source/IcebergSource.java:
```diff
@@ -56,48 +83,85 @@ public boolean supportsExternalMetadata() {
-    // Build Spark table based on Iceberg table, and return it
-    // Eagerly refresh the table before reading to ensure views containing this table show up-to-date data
-    return new SparkTable(icebergTable, schema, true);
+  public Table getTable(StructType schema, Transform[] partitioning, Map<String, String> options) {
```
When implementing `SupportsCatalogOptions`, I don't think this will ever be called. Should we remove it and throw `UnsupportedOperationException` instead?

rymurr replied:
`inferPartitioning` method uses it :-( Can we return `Transform[0]` and have Spark do the work?

rdblue commented on spark3/src/main/java/org/apache/iceberg/spark/source/IcebergSource.java:
```diff
@@ -56,48 +83,85 @@ public boolean supportsExternalMetadata() {
+    SparkSession spark = SparkSession.active();
+    Spark3Util.CatalogAndIdentifier catalogAndIdentifier;
+    try {
+      catalogAndIdentifier = Spark3Util.catalogAndIdentifier(spark, path);
```
Is this try/catch to avoid the `/` check? I think that's a pretty reasonable check. I'm not sure that this is a good idea because the cases you listed above would pass this.

For example, this uses a valid identifier, but is clearly a path reference:
```java
spark.format("iceberg").load("catalog.`file:/path/to/table`");
```

As I mentioned above, I don't think that we should support the mixed path and catalog identifiers, but I still think it is probably a more predictable check to just look for `/`.

rymurr replied:
the try component takes care of `catalog.table`, `catalog.namespace.table` and ```catalog.`file://path/to/table` ``` but doens't take care of `file://path/to/table` which throws a parse exception. The catch part deals w/ `file://path/to/table`.

I suppose we can check 
* if there is a ``` ` ``` character and a `/` and reject (don't mix catalogs and paths)
* If just `/` then treat as path
* assume its tables and parse it

rdblue replied:
Yeah, I think that the problem is that ```catalog.`file://path/to/table` ``` is ambiguous and should not be supported. But parsing it successfully means that it would be translated to a path-based table with a catalog.

I don't think a more complicated check is needed, to see if the `/` is probably escaped. If there is a `/`, then it's a path. That's a really simple rule to follow.

rymurr replied:
After thnking more I have thrown out ``` catalog.`file:/path/to/table` ``` and simplified the method

rdblue commented on spark3/src/main/java/org/apache/iceberg/spark/source/IcebergSource.java:
```diff
@@ -56,48 +83,85 @@ public boolean supportsExternalMetadata() {
+    } catch (ParseException e) {
+      List<String> ident = new ArrayList<>();
+      ident.add(path);
+      catalogAndIdentifier = Spark3Util.catalogAndIdentifier(spark, ident);
```
Nit: Using `ImmutableList` makes this more concise: `Spark3Util.catalogAndIdentifier(spark, ImmutableList.of(path))`

rymurr replied:
fixed

rdblue commented on spark3/src/main/java/org/apache/iceberg/spark/source/IcebergSource.java:
```diff
@@ -56,48 +83,85 @@ public boolean supportsExternalMetadata() {
+    CatalogManager catalogManager = spark.sessionState().catalogManager();
+    String[] currentNamespace = catalogManager.currentNamespace();
+    // we have to check for paths but want to re-use the exiting utils to extract catalog/identifier
+    if (checkPathIdentifier(catalogAndIdentifier.identifier(), currentNamespace)) {
```
I think it would be simpler just to start this method with a check for the path:

```
if (path.contains("/")) {
  Identifier ident = new PathIdentifer(path);
  // do catalog resolution
  return catalog, ident;
}
```

It's easier to document and understand if the rules are simple.

rymurr replied:
agreed, cleaned up

rdblue commented:
I think that the catalog registration here is good. I don't see a way around registering a catalog, so I think we should do it and also use that to maintain compatibility. So here's what I think behavior should be:

1. Always register a default Iceberg catalog that is a HiveCatalog using the HMS URI from hive-site.xml, like what you've done.
2. Use the `/` check for paths. If the table ref is a path, then load it from the default Iceberg catalog (any catalog works).
3. If the table ref has a catalog, use that catalog (if it doesn't support Iceberg, that's fine because we can't guess or replace it)
4. If the table ref does not have a catalog, use the current catalog
5. If the current catalog is the session catalog and is not an `IcebergSessionCatalog`, then replace it with the default Iceberg catalog
6. If the current catalog is not the session catalog, use it even if it doesn't support Iceberg

That keeps behavior identical and also follows Spark's rules for identifiers. The only time that we use the default Iceberg catalog for a Hive identifier is if there is nothing new in Spark 3 that overrides the behavior to use a different catalog, and if the session catalog (equivalent to the default in 2.4) can't handle Iceberg.

Optionally, paths could be loaded from the session catalog if it is an `IcebergSessionCatalog`.

What do you think?

rymurr commented:
Thanks @rdblue I simplified the logic for resolving the default iceberg catalog. A good sign: the change keeps getting smaller :smile:

rdblue commented on site/docs/spark.md:
````diff
@@ -325,6 +325,21 @@ spark.read
+
+*  `file:/path/to/table` -> loads a HadoopTable at given path
+*  ```catalog.`file:/path/to/table` ``` -> fails. Don't set a catalog for paths
+*  ```catalog.namespace.`file:/path/to/table` ``` -> fails. Namespace doesn't exist for paths
````
I think these lines can be removed from the docs. I don't think we need to document cases that would be parsed as strange table identifiers that were never supported.

rymurr replied:
done

rdblue commented on spark3/src/main/java/org/apache/iceberg/spark/source/IcebergSource.java:
```diff
@@ -20,21 +20,45 @@
+ *
+ *  table = "file:/path/to/table" -> loads a HadoopTable at given path
+ *  table = "catalog.`file:/path/to/table`" -> fails. Don't set a catalog for paths
+ *  table = "catalog.namespace.`file:/path/to/table`" -> fails. Namespace doesn't exist for paths
```
Same here, no need for the catalog/path lines.

rymurr replied:
done

rdblue commented on spark3/src/main/java/org/apache/iceberg/spark/source/IcebergSource.java:
```diff
@@ -20,21 +20,45 @@
+ *  table = "catalog.namespace.`file:/path/to/table`" -> fails. Namespace doesn't exist for paths
+ *  table = "tablename" -> loads currentCatalog.currentNamespace.tablename
+ *  table = "catalog.tablename" -> load "tablename" from the specified catalog.
+ *  table = "namespace.tablename" -> Otherwise load "namespace.tablename" from current catalog
```
Nit: starts with "otherwise" still.

rymurr replied:
done

rdblue commented on spark3/src/main/java/org/apache/iceberg/spark/source/IcebergSource.java:
```diff
@@ -56,48 +80,70 @@ public boolean supportsExternalMetadata() {
+      // throwing an iceberg NoSuchTableException because the Spark one is typed and cant be thrown from this interface
+      throw new org.apache.iceberg.exceptions.NoSuchTableException(e, "Cannot find table for %s.", ident);
+    }
+    // throwing an iceberg NoSuchTableException because the Spark one is typed and cant be thrown from this interface
```
Nit: looks like newlines between control flow are missing in a lot of these changes.

rymurr replied:
done.

rdblue commented on spark3/src/main/java/org/apache/iceberg/spark/source/IcebergSource.java:
```diff
@@ -56,48 +80,70 @@ public boolean supportsExternalMetadata() {
-    return new SparkTable(icebergTable, schema, true);
+  public Table getTable(StructType schema, Transform[] partitioning, Map<String, String> options) {
+    String catalogName = extractCatalog(new CaseInsensitiveStringMap(options));
+    Identifier ident = extractIdentifier(new CaseInsensitiveStringMap(options));
```
This is doing a lot of extra work by not calling `CatalogAndIdentifier` directly. There are two maps created, two identical catalog/table resolutions, and then this needs to get the active session and look up the catalog that was already loaded. Is it possible to refactor so that the "extract" functions use a common method that can be used here?

rymurr replied:
yup, this wasn't refactored when I removed some code from the extract methods. Now that they don't do anything I have cleaned up the `getTable` method.

rdblue commented on spark3/src/main/java/org/apache/iceberg/spark/source/IcebergSource.java:
```diff
@@ -56,48 +80,70 @@ public boolean supportsExternalMetadata() {
+          new PathIdentifier(path));
+    }
+    Spark3Util.CatalogAndIdentifier catalogAndIdentifier = Spark3Util.catalogAndIdentifier(spark,
+        DOT.splitToList(path));
```
I think this should parse the identifier using Spark rather than `DOT.splitToList`.

rymurr replied:
done. I have thrown an `IllegalArgumentException` in place of the Spark `ParseException`. This will eventually be caught and turned back into an `AnalysisException`.

rymurr commented:
Rebased and cleaned up based on these comments. Thanks @rdblue 

One thing I wanted to make explicit. The test `TestSparkSchema3` has been removed as `SupportsCatalogOptions` doesn't allow to specify the schema. This introduces a breaking change between Spark2.4 and Spark3 based `IcebergSource`. See https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala#L284

rdblue commented on spark3/src/main/java/org/apache/iceberg/spark/source/IcebergSource.java:
```diff
@@ -56,48 +77,77 @@ public boolean supportsExternalMetadata() {
-        .filter(key -> key.startsWith("hadoop."))
-        .forEach(key -> baseConf.set(key.replaceFirst("hadoop.", ""), options.get(key)));
+  private static void setupDefaultSparkCatalog() {
+    SparkSession spark = SparkSession.active();
```
Sorry for not catching this before, but I think this should be passed in. This should avoid extra calls to `SparkSession.active()`.

rymurr replied:
I should have too. Fixed :-)

rdblue commented on spark3/src/main/java/org/apache/iceberg/spark/source/IcebergSource.java:
```diff
@@ -56,48 +77,77 @@ public boolean supportsExternalMetadata() {
+    try {
+      catalogAndIdentifier = Spark3Util.catalogAndIdentifier(spark, path);
+    } catch (ParseException e) {
+      throw new IllegalArgumentException(String.format("Cannot parse path %s. It is not a valid SQL table", path), e);
```
There's no need to catch `ParseException` any more. A new version of `catalogAndIdentifer` was added with context to form the error message. This could be:

```java
Spark3Util.CatalogAndIdentifier catalogAndIdentifier = Spark3Util.catalogAndIdentifier("identifier", spark, path);
```

rymurr replied:
nice, thanks to @RussellSpitzer! I added another method there to extract the default catalog w/o passing it in.

rdblue commented:
> One thing I wanted to make explicit. The test TestSparkSchema3 has been removed as SupportsCatalogOptions doesn't allow to specify the schema. This introduces a breaking change between Spark2.4 and Spark3 based IcebergSource.

Thanks @rymurr. I think this is the only remaining blocker.

@prodeezy & @xabriel, do you specify the schema when reading Iceberg tables in Spark 3? I think the feature that this PR would disable was added by you, so I want to make sure you're okay with this potential change.

rdblue commented on spark3/src/main/java/org/apache/iceberg/spark/source/IcebergSource.java:
```diff
@@ -56,48 +76,72 @@ public boolean supportsExternalMetadata() {
-    // Overwrite configurations from the Spark Context with configurations from the options.
-    mergeIcebergHadoopConfs(conf, options);
+    final Spark3Util.CatalogAndIdentifier catalogAndIdentifier = Spark3Util.catalogAndIdentifier(
+        "Cannot parse path %s. It is not a valid SQL table", spark, path);
```
The "Cannot parse" part is filled in automatically, so you just need to pass the name of the arg. Probably "path or identifier".

rdblue replied:
That was fast!

rdblue commented:
I think this is ready to go, but we need to build consensus around removing support for custom schemas passed in through `DataFrameWriter`. That was a work-around for not having a better way to project columns in Spark, so hopefully Spark has been fixed in the mean time.

rymurr commented:
> need to build consensus

sure I will send an email to the mailing list. Do you know if there was a specific SPARK Jira related to this?

rdblue commented:
I'm not aware of a JIRA, but it seems like a fairly common use case: needing to supply a specific projection rather than the one that Spark infers.

RussellSpitzer commented on spark3/src/main/java/org/apache/iceberg/spark/source/IcebergSource.java:
```diff
@@ -20,21 +20,41 @@
+ *
+ * How paths/tables are loaded when using spark.read().format("iceberg").path(table)
+ *
+ *  table = "file:/path/to/table" -> loads a HadoopTable at given path
```
Maybe important to note that these are used with this priority as well, ie
if "catalog.namespace.table" is valid it will be read and "namespace.namespace.table" will be ignored ?

rymurr replied:
very good point @RussellSpitzer, I have included a note to that effect

xabriel commented:
> > One thing I wanted to make explicit. The test TestSparkSchema3 has been removed as SupportsCatalogOptions doesn't allow to specify the schema. This introduces a breaking change between Spark2.4 and Spark3 based IcebergSource.
> 
> Thanks @rymurr. I think this is the only remaining blocker.
> 
> @prodeezy & @xabriel, do you specify the schema when reading Iceberg tables in Spark 3? I think the feature that this PR would disable was added by you, so I want to make sure you're okay with this potential change.

Thank you for the ping @rdblue.

All of our current code is based on Spark 2.4.X branch. We meant to pursue this further and to suggest a new API on the Spark 3 branch, but other work took precedence. The discussion is over at #590 (TL;DR: seems like we should have a `df.project(StructType)` API).

So yes this change will break us once we move to Spark 3. However I want to discuss internally whether we feel strongly about it or not. Will get back to this thread ASAP.

rymurr commented:
Hey @xabriel any thoughts or feedback on this change?

xabriel commented:
@rymurr from our point of view, go for it. We are more interested in the custom catalog support you're building than the user `schema` feature.

( Sorry it took me so long to reply. )

rymurr commented:
> @rymurr from our point of view, go for it. We are more interested in the custom catalog support you're building than the user `schema` feature.
> 
> ( Sorry it took me so long to reply. )

Rock on! Thanks @xabriel

rdblue commented:
Thanks for checking, @xabriel!

I'll merge this when tests are passing.

rdblue commented:
Thanks for all your hard work on this, @rymurr! Great to have this working!
